src/librustdoc/json/mod.rs

```diff
@@ -264,20 +278,21 @@ impl<'tcx> FormatRenderer<'tcx> for JsonRenderer<'tcx> {
                 .collect(),
             format_version: types::FORMAT_VERSION,
         };
-        let out_dir = self.out_path.clone();
-        try_err!(create_dir_all(&out_dir), out_dir);
+        if let Some(ref out_path) = self.out_path {
+            let out_dir = out_path.clone();
+            try_err!(create_dir_all(&out_dir), out_dir);
 
-        let mut p = out_dir;
-        p.push(output.index.get(&output.root).unwrap().name.clone().unwrap());
-        p.set_extension("json");
-        let mut file = BufWriter::new(try_err!(File::create(&p), p));
-        self.tcx
-            .sess
-            .time("rustdoc_json_serialization", || serde_json::ser::to_writer(&mut file, &output))
-            .unwrap();
-        try_err!(file.flush(), p);
-
-        Ok(())
+            let mut p = out_dir;
+            p.push(output.index.get(&output.root).unwrap().name.clone().unwrap());
```

Comment:
We should also respect the `--output` flag here, even if it's not for a file. Currently `--output` is ignored for JSON, but if `--output -` works, then `--output some_file.json` should also work to rename the JSON (instead of using the crate name).

Reply:
Good idea!

Reply:
So, from what I can see, `--out-dir` and `--output` are kinda considered the same by rustdoc and `--output` is deprecated. So currently we don't really have a way to say I want JSON output to be generated into "this file".

Reply:
Remark (peanut gallery): I would've thought `--output=PATH` would cause the json output to be dumped to this *specific* file at `PATH`, and that if `--output` is specified, `--out-dir` would be ignored.

Reply:
Yes it's the whole problem here. The `--output` option was created when the HTML format was the only format, making it useless and why we created `--out-dir`. Currently, both options are actually the same, which is pretty bad.